includes/class-wcs-action-scheduler-customer-notifications.php

```diff
@@ -0,0 +1,290 @@
+
+		//TODO: For end of trial, should we schedule payment notification? Let's say no.
+		if ( $subscription->get_date( 'trial_end' ) ) {
+			return;
+		}
+
+		$next_payment = $subscription->get_date( 'next_payment' );
+		$timestamp    = $this->sub_time_offset( $next_payment, $subscription );
+
+		// Can manual vs automatic payment change until it runs?
+		if ( $subscription->is_manual() ) {
+			$this->schedule_notification(
+				$subscription,
+				'woocommerce_scheduled_subscription_customer_notification_manual_renewal',
+				$timestamp
+			);
+		} else {
+			$this->schedule_notification(
+				$subscription,
+				'woocommerce_scheduled_subscription_customer_notification_auto_renewal',
+				$timestamp
+			);
+		}
+	}
+
+	/**
+	 * Maybe set a schedule action if the new date is in the future
+	 *
+	 * @param object $subscription An instance of a WC_Subscription object
+	 * @param string $date_type Can be 'trial_end', 'next_payment', 'payment_retry', 'end', 'end_of_prepaid_term' or a custom date type
+	 * @param string $datetime A MySQL formatted date/time string in the GMT/UTC timezone.
+	 */
+	public function update_date( $subscription, $date_type, $datetime ) {
+
+		if ( in_array( $date_type, $this->get_date_types_to_schedule(), true ) ) {
+
+			if ( $subscription->get_date( 'trial_end' ) ) {
+				$this->schedule_trial_ending_notification( $subscription );
+			}
+
+			if ( $subscription->get_date( 'end' ) ) {
+				$this->schedule_expiry_notification( $subscription );
+			}
+
+			if ( $subscription->get_date( 'next_payment' ) ) {
+				$this->schedule_payment_notification( $subscription );
+			}
+		}
+	}
+
+	/**
+	 * Delete a date from the action scheduler queue
+	 *
+	 * @param object $subscription An instance of a WC_Subscription object
+	 * @param string $date_type Can be 'trial_end', 'next_payment', 'end', 'end_of_prepaid_term' or a custom date type
+	 */
+	public function delete_date( $subscription, $date_type ) {
+		$this->update_date( $subscription, $date_type, '0' );
+	}
+
+	/**
+	 * When a subscription's status is updated, maybe schedule an event
+	 *
+	 * @param object $subscription An instance of a WC_Subscription object
+	 * @param string $date_type Can be 'trial_end', 'next_payment', 'end', 'end_of_prepaid_term' or a custom date type
+	 * @param string $datetime A MySQL formatted date/time string in the GMT/UTC timezone.
+	 */
+	public function update_status( $subscription, $new_status, $old_status ) {
+
+		switch ( $new_status ) {
+			case 'active':
+				if ( $subscription->get_date( 'trial_end' ) ) {
+					$this->schedule_trial_ending_notification( $subscription );
+				}
+
+				if ( $subscription->get_date( 'end' ) ) {
+					$this->schedule_expiry_notification( $subscription );
+				}
+
+				if ( $subscription->get_date( 'next_payment' ) ) {
+					$this->schedule_payment_notification( $subscription );
+				}
+
+				break;
+			case 'pending-cancel':
+				// Unschedule all notifications?
```

Comment:
I don't think you would want to unschedule all, I think the expiration one is still one you'd want.

Reply:
Agree, should be resolved now.